typescript/core/src/McapStreamReader.ts

```diff
@@ -89,7 +91,61 @@ export default class McapStreamReader {
     if (this.#doneReading) {
       throw new Error("Already done reading");
     }
-    this.#buffer.append(data);
+    this.#appendOrShift(data);
+  }
+  #appendOrShift(data: Uint8Array): void {
+    /** Add data to the buffer, shifting existing data or reallocating if necessary. */
+    const consumedBytes = this.#reader.offset;
+    const remainingBytes = this.#view.byteLength - consumedBytes;
+    const totalNeededBytes = remainingBytes + data.byteLength;
```

Comment:
How is `totalNeededBytes` correct here? `remainingBytes` is what has not been used in the buffer and then `data.byteLength` is the size of the incoming data. Does `totalNeededBytes` represent the amount of bytes we _need_ to fit the data? I don't understand what it is meant to represent.

This also makes it hard to answer whether the _if_ statement on the next line is correct. It compares `totalNeededBytes` to the length of the buffer which assumes that totalNeededBytes is how much is needed to store existing + new data but it doesn't seem obvious to me that `totalNeededBytes` includes existing data?

Reply:
After initially writing a comment in support of roman's statement, I was about to press the submit button but then I double-checked and now I think I actually understand what is happening here — but I think we should find a better name for `remainingBytes`. It's not the amount of _capacity remaining_ in the buffer, but it's actually the number of _stored bytes that have not been consumed yet_. I think a better name would be `unconsumedBytes` (to contrast with `consumedBytes`) or `unreadBytes`.

The `totalNeededBytes` are the number of bytes we will need to store – this includes the ones which haven't been consumed yet + the new ones in `data`.  Maybe that should be named `neededCapacity`?